source/anonymous-data-logger/lib/metrics.py

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+##############################################################################
+#  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   #
+#                                                                            #
+#  Licensed under the Apache License Version 2.0 (the "License").            #
+#  You may not use this file except in compliance with the License.          #
+#  A copy of the License is located at                                       #
+#                                                                            #
+#      http://www.apache.org/licenses/                                       #
+#                                                                            #
+#  or in the "license" file accompanying this file. This file is distributed #
+#  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        #
+#  express or implied. See the License for the specific language governing   #
+#  permissions and limitations under the License.                            #
+##############################################################################
+
+import datetime
+import json
+import urllib.request
+
+def send_metrics(config):
+    metrics = {}
+    # move Solution ID and UUID to the root JSON level
+    metrics['Solution'] = config.pop("SolutionId", None)
+    metrics['UUID'] = config.pop("UUID", None)
+    metrics['TimeStamp'] = str(datetime.datetime.utcnow().isoformat())
+    metrics['Data'] = config
+    url = 'https://metrics.awssolutionsbuilder.com/generic'
+    data = json.dumps(metrics).encode('utf8')
+    headers = {'content-type': 'application/json'}
+    req = urllib.request.Request(url, data,headers)
+    response = urllib.request.urlopen(req)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *response*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close *response* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).

Reply:
We are not going to address this CodeGuru finding because we want to mirror the official AWS code sample for sigv4 signing: https://docs.aws.amazon.com/code-samples/latest/catalog/python-signv4-v4-signing-get-post.py.html.